fastn-core/src/auth/emailpassword.rs

```diff
@@ -0,0 +1,371 @@
+        // TODO: use correct HTTP status code
+        return fastn_core::http::api_error("user already exists");
+    }
+
+    // encrypt password
+    let salt =
+        argon2::password_hash::SaltString::generate(&mut argon2::password_hash::rand_core::OsRng);
+
+    let argon2 = argon2::Argon2::default();
+
+    let hashed_password = argon2
+        .hash_password(user.password.as_bytes(), &salt)
+        .unwrap()
+        .to_string();
+
+    let user_id = uuid::Uuid::new_v4();
+
+    // TODO: email verification
+    // TODO: password can be null for oauth users. if it's null they should not login using
+    // emailpassword flow
+    let user = upsert_user(
+        &user_id,
+        user.email.as_str(),
+        user.username.as_str(),
+        user.name.as_str(),
+        hashed_password.as_str(),
+    )
+    .await?;
+
+    tracing::info!("fastn_user created. user: {:?}", &user);
+
+    fastn_core::http::api_ok(user)
+}
+
+pub(crate) async fn login(
+    req: &crate::http::Request,
+    next: String,
+) -> fastn_core::Result<fastn_core::http::Response> {
+    if req.method() != "POST" {
+        return Ok(fastn_core::not_found!("invalid route"));
+    }
+
+    #[derive(serde::Deserialize, Debug)]
+    struct Payload {
+        // TODO: add email/username
+        email: String,
+        password: String,
+    }
+
+    let body: Payload = serde_json::from_slice(req.body())?;
+
+    tracing::info!("request payload: {:?}", body);
+
+    let client = db::get_client().await?;
+
+    let user = client
+        .query(
+            "select id, username, email, name, password from fastn_user where email = $1 limit 1",
+            &[&body.email],
+        )
+        .await
+        .unwrap_or(vec![]);
+
+    if user.is_empty() {
+        return fastn_core::http::api_error("invalid payload");
+    }
+
+    let user = user.first().expect("vec of length must have first");
+
+    let password: String = user.get("password");
+    let user_id: uuid::Uuid = user.get("id");
+
+    let parsed_hash = argon2::PasswordHash::new(&password).unwrap();
+
+    let password_match = Argon2::default().verify_password(body.password.as_bytes(), &parsed_hash);
+
+    if password_match.is_err() {
+        return fastn_core::http::api_error("invalid payload");
+    }
+
+    let session_id = uuid::Uuid::new_v4();
+
+    let affected = create_session(&session_id, &user_id).await.unwrap_or(0);
+
+    tracing::info!("session created. Rows affected: {}", &affected);
+
+    // client has to 'follow' this request
+    // https://stackoverflow.com/a/39739894
+    fastn_core::auth::set_session_cookie_and_end_response(req, session_id, next).await
+}
+
+/// insert user if it not exists in the database returning the rows affected
+/// email and username is unique in fastn_user table
+pub async fn upsert_user(
+    user_id: &uuid::Uuid,
+    email: &str,
+    username: &str,
+    name: &str,
+    hashed_password: &str,
+) -> fastn_core::Result<fastn_core::auth::FastnUser> {
```

Comment:
I am not fond of this kind of functions. At every call site you know precisely if the user must exist and it must be an update user call, or user must not exist and it is insert call. You want things to fail as soon as possible, if code has a bug it must crash the server tonight, instead of silently corrupting data for years.